listenbrainz/webserver/static/js/src/personal-recommendations/PersonalRecommendationsModal.tsx

```diff
@@ -0,0 +1,263 @@
+      this.setState({ suggestions: [] });
+    }
+  };
+
+  submitPersonalRecommendation = async () => {
+    const { recordingToPersonallyRecommend, newAlert } = this.props;
+    const { blurbContent, users } = this.state;
+    const { APIService, currentUser } = this.context;
+
+    if (currentUser?.auth_token) {
+      const metadata: UserTrackPersonalRecommendationMetadata = {
+        artist_name: getArtistName(recordingToPersonallyRecommend),
+        track_name: getTrackName(recordingToPersonallyRecommend),
+        release_name: recordingToPersonallyRecommend!.track_metadata
+          .release_name,
+        recording_mbid: getRecordingMBID(recordingToPersonallyRecommend!),
+        recording_msid: getRecordingMSID(recordingToPersonallyRecommend!),
+        users,
+        blurb_content: blurbContent,
+      };
+      try {
+        const status = await APIService.submitPersonalRecommendation(
+          currentUser.auth_token,
+          currentUser.name,
+          metadata
+        );
+        if (status === 200) {
+          newAlert(
+            "success",
+            "You personally recommended a track!",
+            `${metadata.artist_name} - ${metadata.track_name}`
+          );
+          this.setState({ blurbContent: "" });
+        }
+      } catch (error) {
+        this.handleError(error, "Error while personally recommending");
+      }
+    }
+  };
+
+  render() {
+    const { recordingToPersonallyRecommend } = this.props;
+    if (!recordingToPersonallyRecommend) {
+      return null;
+    }
+    const { blurbContent, users, suggestions } = this.state;
+    const { track_name } = recordingToPersonallyRecommend.track_metadata;
+    const { artist_name } = recordingToPersonallyRecommend.track_metadata;
+    const { APIService, currentUser } = this.context;
+    return (
+      <div
+        className="modal fade"
+        id="PersonalRecommendationModal"
+        tabIndex={-1}
+        role="dialog"
+        aria-labelledby="PersonalRecommendationModalLabel"
+        data-backdrop="static"
+      >
+        <div className="modal-dialog" role="document">
+          <form className="modal-content">
+            <div className="modal-header">
+              <button
+                type="button"
+                className="close"
+                data-dismiss="modal"
+                aria-label="Close"
+                onClick={() => {
+                  this.setState({
+                    users: [],
+                    blurbContent: "",
+                    suggestions: [],
+                  });
+                  // update the followers list
+                  APIService.getFollowersOfUser(currentUser.name)
+                    .then((response) => {
+                      this.setState({ followers: response.followers });
+                    })
+                    .catch((error) => {
+                      this.handleError(error, "Error while fetching followers");
+                    });
+                }}
```

Comment:
It's best no to have anonymous functions defined in the render method, because they will get re-created on every re-render (in theory that's not optimal).

What you want is to have a method defined on the class such as `onCloseModal` that will call setState.
You will also want to refactor the code to get followers that you have in `componentDidMount` into its own method `getFollowers`, so that you can call it from componentDidMount and from onCloseModal without having duplicate code.

P.S: why do we need to update the followers list when we close the modal? These two things don't seem like they should be related

Reply:
Makes sense, thanks

> P.S: why do we need to update the followers list when we close the modal? These two things don't seem like they should be related

I did it mostly for the UX
for example, when someone sends you an email, you would refresh until you get the email. Similarly, suppose you are recommending someone, and you wrote a huge recommendation, only to realise that someone doesn't follow you, and you make ask that person to follow you. But sadly you would have to reload the whole page to have an updated list.

Reply:
I think that's not a bad idea, but maybe not the right design for that.
Without a clear indication I wouldn't suspect that the list would refresh automatically.
Maybe a well-placed refresh button might be helpful for that? I wonder how much use it will get, but at the same time if it just calls a function that is already implemented it doesn't cost anything to add it.